_posts/2022-12-31-arrow-adbc.md

```diff
@@ -0,0 +1,217 @@
+So generally, columnar data must be converted to rows between steps 5 and 6, spending resources without performing "useful" work.
+
+This mismatch is problematic for columnar database systems, such as ClickHouse, Dremio, DuckDB, and Google BigQuery.
+On the client side, tools such as Apache Spark and pandas would be better off getting columnar data directly, skipping that conversion.
+Otherwise, they're leaving performance on the table.
+At the same time, that conversion isn't always avoidable.
+Row-oriented database systems like PostgreSQL aren't going away, and these clients will still want to consume data from them.
+
+Developers have a few options:
+
+- *Just use JDBC/ODBC*.
+  These standards are here to stay, and it makes sense for databases to support them for applications that want them.
+  But when both the database and the application are columnar, that means converting data into rows for JDBC/ODBC, only for the client to convert them right back into columns!
+  Performance suffers, and developers have to spend time implementing the conversions.
+- *Use JDBC/ODBC to Arrow conversion libraries*.
+  Libraries like [Turbodbc][turbodbc] and [arrow-jdbc][arrow-jdbc] handle row to columnar conversions for clients.
+  But this doesn't fundamentally solve the problem.
+  Unnecessary data conversions are still required.
+- *Use vendor-specific protocols*.
+  For some databases, applications can use a database-specific protocol or SDK to directly get Arrow data.
+  For example, applications could use Dremio via [Arrow Flight SQL][flight-sql].
+  But client applications that want to use multiple database vendors would need to integrate with each of them.
+  (Look at all the [connectors](https://trino.io/docs/current/connector.html) that Trino implements.)
+  And databases like PostgreSQL don't offer an option supporting Arrow in the first place.
+
+So in the status quo, clients must choose between either tedious integration work or leaving performance on the table.
+
+## Introducing ADBC
+
+ADBC is an Arrow-based, vendor-netural API for interacting with databases.
+Applications that use ADBC just get Arrow data.
+They don't have to do any conversions themselves, and they don't have to integrate each database's specific SDK.
+
+Just like JDBC/ODBC, underneath the ADBC API are drivers that translate the API for specific databases.
+
+* A driver for an Arrow-native database just passes Arrow data through without conversion.
+* A driver for a non-Arrow-native database must convert the data to Arrow.
+  This saves the application from doing that, and the driver can optimize the conversion for its database.
+
+<figure style="text-align: center;">
+  <img src="{{ site.baseurl }}/img/ADBCFlow2.svg" alt="A diagram showing the query execution flow with ADBC." width="90%" class="img-responsive">
+  <figcaption>The query execution flow with two different ADBC drivers.</figcaption>
+</figure>
+
```

Comment:
Here the numbered list follows the diagram, whereas above, the numbered list precedes the diagram. It'd probably be best to have the order the same in both places.

Reply:
Updated (diagram precedes list in both places)